services/venice-controller/src/test/java/com/linkedin/venice/controller/kafka/protocol/admin/SemanticDetectorTest.java

```diff
@@ -0,0 +1,530 @@
+        + "      \"type\": \"enum\", \"name\": \"ExecutionType\", \"symbols\": [\"START\", \"COMPLETED\"]}" + "    }"
+        + "  ]" + "}" + "}";
+
+    Schema currentSchema = AvroCompatibilityHelper.parse(schemaJson);
+    Schema targetSchema = AvroCompatibilityHelper.parse(targetSchemaJson);
+
+    // Create records for the schema
+    Schema recordSchema = currentSchema.getElementType();
+    GenericRecord record1 = new GenericData.Record(recordSchema);
+    record1.put("field1", "exampleString");
+    record1.put("field2", 123);
+    record1.put("executionType", "START");
+
+    GenericRecord record2 = new GenericData.Record(recordSchema);
+    record2.put("field1", "");
+    record2.put("field2", 0);
+    record2.put("executionType", "STOP");
+
+    GenericRecord record3 = new GenericData.Record(recordSchema);
+    record3.put("field1", "");
+    record3.put("field2", 0);
+    record3.put("executionType", "COMPLETED");
+
+    // Create an array and add the record to it
+    GenericData.Array<GenericRecord> array1 = new GenericData.Array<>(2, currentSchema);
+    array1.add(record1);
+    array1.add(record2);
+
+    testBadSemanticUsage(
+        array1,
+        currentSchema,
+        "",
+        targetSchema,
+        "Field array.1.ExampleRecord.executionType: Enum value STOP is not accepted in the target schema [START, COMPLETED]");
+
+    GenericData.Array<GenericRecord> array2 = new GenericData.Array<>(2, currentSchema);
+    array2.add(record1);
+    array2.add(record3);
+
+    testBadSemanticUsage(
+        array2,
+        currentSchema,
+        "",
+        targetSchema,
+        "Field array.1.ExampleRecord.executionType: Invalid enum value COMPLETED is not accepted in the current schema [START, STOP, PAUSE]");
+  }
+
+  @Test
+  public void testValidateEnum() {
+    Schema currentSchema = AvroCompatibilityHelper.newEnumSchema(
+        "currentEnum",
+        "",
+        "NewSemanticUsageValidatorTest",
+        Arrays.asList("value1", "value2", "value3", "value4"),
+        null);
+
+    Schema targetSchema = AvroCompatibilityHelper.newEnumSchema(
+        "targetEnum",
+        "",
+        "NewSemanticUsageValidatorTest",
+        Arrays.asList("value1", "value2", "value3"),
+        null);
+
+    SemanticDetector.validateEnum("value1", currentSchema, targetSchema, "fieldA");
+
+    testBadSemanticUsage(
+        "value4",
+        currentSchema,
+        "fieldA",
+        targetSchema,
+        "Field fieldA: Enum value value4 is not accepted in the target schema [value1, value2, value3]");
+
+    testBadSemanticUsage(
+        "value5",
+        currentSchema,
+        "fieldA",
+        targetSchema,
+        "Field fieldA: Invalid enum value value5 is not accepted in the current schema [value1, value2, value3, value4]");
+
+    testBadSemanticUsage(1, currentSchema, "fieldB", targetSchema, "Field fieldB: Enum value 1 is not a string");
+  }
+
+  @Test
+  public void testFindObjectSchemaInsideUnion() {
```

Comment:
Could you please add a bad semantic usage test for union schema? For example, target schema only has one record schema in the union, while the current schema has two different record schemas in the union; we test that traverse would fail if we start using the second record in the union.

Reply:
Added.